plugin/remotecall/main.go

```diff
@@ -0,0 +1,139 @@
+}
+
+func generateUpstreamUserName(response *remotecall.UserKeyAuthResponse) string {
+	return "user" + "." + response.Data.UUID
+}
+
+const (
+	remoteAuthEndpoints       = "remote-auth-endpoint"
+	remoteAuthEndpointsSecret = "remote-auth-endpoint-secret"
+	remoteEndpoints           = "remote-endpoint"
+	userClusterEndpoint       = "user-cluster-endpoint"
+	userClusterEndpointToken  = "user-cluster-endpoint-token"
+	mappingKeyPath            = "mapping-key-path"
+)
+
+func main() {
+	libplugin.CreateAndRunPluginTemplate(&libplugin.PluginTemplate{
+		Name:  "remote call",
+		Usage: "sshpiperd remote plugin",
+		Flags: []cli.Flag{
+			&cli.GenericFlag{
+				Name:     remoteAuthEndpoints,
+				Usage:    "cluster-url map for remote endpoint for retrieving user's private key(given as prod1=url)",
+				EnvVars:  []string{"SSHPIPERD_PRIVATE_KEY_ENDPOINTS"},
+				Value:    &remotecall.StringMapFlag{},
+				Required: true,
+			},
+			&cli.GenericFlag{
+				Name:     remoteAuthEndpointsSecret,
+				Usage:    "cluster-secret map for cluster-url for auth(given as prod1=token)",
+				EnvVars:  []string{"SSHPIPERD_PRIVATE_KEY_ENDPOINTS"},
+				Value:    &remotecall.StringMapFlag{},
+				Required: true,
+			},
+			&cli.GenericFlag{
+				Name:     remoteEndpoints,
+				Usage:    "path to remote endpoint for forwarding traffic",
+				EnvVars:  []string{"SSHPIPERD_IN_CLUSTER_ENDPOINTS"},
+				Value:    &remotecall.StringMapFlag{},
+				Required: true,
+			},
+			&cli.StringFlag{
+				Name:    userClusterEndpoint,
+				Usage:   "endpoint for getting user to cluster mapping",
+				EnvVars: []string{"SSHPIPERD_USER_MAPPING_ENDPOINT"},
+			},
+			&cli.StringFlag{
+				Name:    userClusterEndpointToken,
+				Usage:   "auth token(added to header) for getting user to cluster mapping",
+				EnvVars: []string{"SSHPIPERD_USER_MAPPING_ENDPOINT_TOKEN"},
+			},
+			&cli.PathFlag{
+				Name:    mappingKeyPath,
+				Usage:   "mapping key for upstream",
+				EnvVars: []string{"SSHPIPERD_MAPPING_KEY_PATH"},
+			},
+		},
+		CreateConfig: func(c *cli.Context) (*libplugin.SshPiperPluginConfig, error) {
+			return createConfig(c)
+		},
+	})
+}
+
+func createConfig(c *cli.Context) (*libplugin.SshPiperPluginConfig, error) {
+	caller, err := createRemoteCaller(c)
+	if err != nil {
+		return nil, fmt.Errorf("error creating remote caller: %w", err)
+	}
+
+	return &libplugin.SshPiperPluginConfig{
+		PublicKeyCallbackNew: func(conn libplugin.ConnMetadata, key []byte, keytype string) (*libplugin.Upstream, error) {
+			clusterName, err := caller.GetClusterName(conn.User())
+			log.Debugf("username %s", conn.User())
+			if err != nil {
+				return nil, fmt.Errorf("error getting cluster name from user: %w", err)
+			}
+
+			clusterAuthnURL, err := caller.GetUpstreamAuthenticatorURL(clusterName)
+			if err != nil {
+				return nil, fmt.Errorf("error getting authenticator url from cluster name: %w", err)
+			}
+			clusterAuthnToken, err := caller.GetUpstreamAuthenticatorAuthToken(clusterName)
```

Comment:
is there a reason caller is exposing `GetUpstreamAuthenticatorURL` and `GetUpstreamAuthenticatorAuthToken` just so we provide it back as part of `AuthenticateKey` method? or could we just say authenticate key and provide `keyType`, `clustername`, and `conn.User`?